File: src/lib/types/import.ts
/**
 * This module defines typed front-end contracts for import inspection, batch review, and rollback follow-through.
 *
 * Why this file exists:
 * - The UI reads these shapes as its desktop and preview contract, so unclear names here ripple through every consumer.
 * - If you need to know what a route or helper expects from the backend, this is often the fastest file to open first.
 *
 * Main declarations:
 * - `TakeoutRequest`
 * - `TakeoutFileReport`
 * - `TakeoutPreviewEntry`
 * - `ImportBatchOverview`
 * - `ImportBatchDetail`
 * - `TakeoutInspection`
 *
 * Source-of-truth notes:
 * - Data shapes should stay aligned with the accepted architecture and feature docs rather than ad-hoc page assumptions.
 * - Prefer additive, explicit fields over ambiguous catch-all objects so the trust surface stays auditable.
 */

/**
 * Describes a request payload in this front-end contract.
 *
 * These type contracts are read directly by routes, helper modules, and preview fixtures, so a reader should be able to understand the shape without hunting through call sites.
 */
export interface TakeoutRequest {
  sourcePath: string
  dryRun: boolean
}

/**
 * Describes a browser-direct import request.
 *
 * Browser Direct uses the same preview/import review models as Takeout, but it
 * carries browser profile metadata so the backend can preserve the source
 * profile instead of routing a local SQLite database through the Takeout parser.
 */
export interface BrowserHistoryImportRequest {
  sourcePath: string
  dryRun: boolean
  browserFamily?: string | null
  profileId?: string | null
  browserName?: string | null
  profileName?: string | null
}

/**
 * Represents a completed report that the UI can review after a run finishes.
 *
 * These type contracts are read directly by routes, helper modules, and preview fixtures, so a reader should be able to understand the shape without hunting through call sites.
 */
export interface TakeoutFileReport {
  path: string
  kind: string
  status: string
  records: number
  classification: string
  reasonCode?: string | null
  reasonDetail?: string | null
  detectedLocale?: string | null
}

/**
 * Defines the typed shape for takeout preview entry.
 *
 * These type contracts are read directly by routes, helper modules, and preview fixtures, so a reader should be able to understand the shape without hunting through call sites.
 */
export interface TakeoutPreviewEntry {
  sourcePath: string
  url: string
  title?: string | null
  visitedAt: string
  sourceVisitId: number
  status: string
}

/**
 * Defines the typed shape for import batch overview.
 *
 * These type contracts are read directly by routes, helper modules, and preview fixtures, so a reader should be able to understand the shape without hunting through call sites.
 */
export interface ImportBatchOverview {
  id: number
  sourceKind: string
  sourcePath: string
  profileId: string
  createdAt: string
  importedAt?: string | null
  revertedAt?: string | null
  status: string
  candidateItems: number
  importedItems: number
  duplicateItems: number
  visibleItems: number
  auditPath?: string | null
  gitCommit?: string | null
}

/**
 * Represents the detailed view model for import batch.
 *
 * These type contracts are read directly by routes, helper modules, and preview fixtures, so a reader should be able to understand the shape without hunting through call sites.
 */
export interface ImportBatchDetail {
  batch: ImportBatchOverview
  previewEntries: TakeoutPreviewEntry[]
  recognizedFiles: TakeoutFileReport[]
  quarantinedFiles: TakeoutFileReport[]
  notes: string[]
  detectedLocale?: string | null
  previewRangeStart?: string | null
  previewRangeEnd?: string | null
}

/**
 * Defines the typed shape for takeout inspection.
 *
 * These type contracts are read directly by routes, helper modules, and preview fixtures, so a reader should be able to understand the shape without hunting through call sites.
 */
export interface TakeoutInspection {
  dryRun: boolean
  sourcePath: string
  recognizedFiles: TakeoutFileReport[]
  quarantinedFiles: TakeoutFileReport[]
  previewEntries: TakeoutPreviewEntry[]
  candidateItems: number
  importedItems: number
  duplicateItems: number
  notes: string[]
  importBatch?: ImportBatchOverview | null
  detectedLocale?: string | null
  previewRangeStart?: string | null
  previewRangeEnd?: string | null
}

/**
 * Defines the typed shape for import progress event.
 *
 * These type contracts are read directly by routes, helper modules, and preview fixtures, so a reader should be able to understand the shape without hunting through call sites.
 */
export interface ImportProgressEvent {
  phase: string
  label: string
  detail: string
  current: number
  total: number
  progressPercent?: number | null
  logLines: string[]
  sourcePath?: string | null
  sourceLabel?: string | null
  processedRecords?: number | null
  totalRecords?: number | null
  importedRecords?: number | null
  duplicateRecords?: number | null
  skippedRecords?: number | null
}

File: src-tauri/crates/vault-core/src/models/import.rs
//! Takeout and import-batch read models.

use serde::{Deserialize, Serialize};

/// Request payload for inspecting or importing a Takeout source.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TakeoutRequest {
    pub source_path: String,
    pub dry_run: bool,
}

/// Summary of one recognized or quarantined file inside a Takeout source.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct TakeoutFileReport {
    pub path: String,
    pub kind: String,
    pub status: String,
    pub records: usize,
    pub classification: String,
    pub reason_code: Option<String>,
    pub reason_detail: Option<String>,
    pub detected_locale: Option<String>,
}

/// One preview visit shown before or after a Takeout import.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct TakeoutPreviewEntry {
    pub source_path: String,
    pub url: String,
    pub title: Option<String>,
    pub visited_at: String,
    pub source_visit_id: i64,
    pub status: String,
}

/// Compact summary for one recorded import batch.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ImportBatchOverview {
    pub id: i64,
    pub source_kind: String,
    pub source_path: String,
    pub profile_id: String,
    pub created_at: String,
    pub imported_at: Option<String>,
    pub reverted_at: Option<String>,
    pub status: String,
    pub candidate_items: usize,
    pub imported_items: usize,
    pub duplicate_items: usize,
    pub visible_items: usize,
    pub audit_path: Option<String>,
    pub git_commit: Option<String>,
}

/// Detailed import-batch view with preview rows and file reports.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ImportBatchDetail {
    pub batch: ImportBatchOverview,
    pub preview_entries: Vec<TakeoutPreviewEntry>,
    pub recognized_files: Vec<TakeoutFileReport>,
    pub quarantined_files: Vec<TakeoutFileReport>,
    pub notes: Vec<String>,
    pub detected_locale: Option<String>,
    pub preview_range_start: Option<String>,
    pub preview_range_end: Option<String>,
}

/// Full inspection/import payload returned by the Takeout flow.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct TakeoutInspection {
    pub source_path: String,
    pub dry_run: bool,
    pub recognized_files: Vec<TakeoutFileReport>,
    pub quarantined_files: Vec<TakeoutFileReport>,
    pub candidate_items: usize,
    pub imported_items: usize,
    pub duplicate_items: usize,
    pub preview_entries: Vec<TakeoutPreviewEntry>,
    pub import_batch: Option<ImportBatchOverview>,
    pub notes: Vec<String>,
    pub detected_locale: Option<String>,
    pub preview_range_start: Option<String>,
    pub preview_range_end: Option<String>,
}

/// Progress event streamed while a Takeout import is running.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ImportProgressEvent {
    pub phase: String,
    pub label: String,
    pub detail: String,
    pub current: usize,
    pub total: usize,
    pub progress_percent: Option<f32>,
    pub log_lines: Vec<String>,
    pub source_path: Option<String>,
}
